packages/convex-helpers/server/pagination.ts

```diff
@@ -0,0 +1,293 @@
+  /// Default is 100.
+  targetMaxRows?: number;
+  /// Absolute maximum number of rows to return, even if endIndexKey is
+  /// provided. Use this to prevent a single page from growing too large, but
+  /// watch out because gaps can form between pages.
+  /// Default is unlimited.
+  absoluteMaxRows?: number;
+  /// Whether the index is walked in ascending or descending order. Default is
+  /// ascending.
+  order?: "asc" | "desc";
+  /// Which index to walk.
+  /// Default is by_creation_time.
+  index?: IndexNames<NamedTableInfo<DataModel, T>>;
+  /// If index is not by_creation_time or by_id,
+  /// you need to provide the index fields, either directly or from the schema.
+  /// schema can be found with
+  /// `import schema from "./schema";`
+  schema?: any;
+  indexFields?: string[];
+};
+
+export type PageResponse<
+  DataModel extends GenericDataModel,
+  T extends TableNamesInDataModel<DataModel>,
+> = {
+  page: DocumentByName<DataModel, T>[];
+  hasMore: boolean;
+  // indexKeys[i] is the index key for the document page[i].
+  indexKeys: IndexKey[];
+};
+
+/// Get a single page of documents from a table.
+/// See examples in README.
+export async function getPage<
+  DataModel extends GenericDataModel,
+  T extends TableNamesInDataModel<DataModel>,
+>(
+  ctx: GenericQueryCtx<DataModel>,
+  request: PageRequest<DataModel, T>,
+): Promise<PageResponse<DataModel, T>> {
+  const index = request.index ?? "by_creation_time";
+  const indexFields = getIndexFields(request);
+  const startIndexKey = request.startIndexKey ?? [];
+  const endIndexKey = request.endIndexKey ?? [];
+  const startInclusive = request.startInclusive ?? false;
+  const order = request.order === "desc" ? "desc" : "asc";
+  const startBoundType =
+    order === "desc" ? ltOr(startInclusive) : gtOr(startInclusive);
+  const endInclusive = request.endInclusive ?? true;
+  const endBoundType =
+    order === "desc" ? gtOr(endInclusive) : ltOr(endInclusive);
+  if (
+    indexFields.length < startIndexKey.length ||
+    indexFields.length < endIndexKey.length
+  ) {
+    throw new Error("Index key length exceeds index fields length");
+  }
+  const split = splitRange(
+    indexFields,
+    startIndexKey,
+    endIndexKey,
+    startBoundType,
+    endBoundType,
+  );
+  const absoluteMaxRows = request.absoluteMaxRows ?? Infinity;
+  const targetMaxRows = request.targetMaxRows ?? DEFAULT_TARGET_MAX_ROWS;
+  const absoluteLimit = request.endIndexKey
+    ? absoluteMaxRows
+    : Math.min(absoluteMaxRows, targetMaxRows);
+  const page = [];
+  const indexKeys = [];
+  let hasMore = false;
+  for (const range of split) {
+    const query = ctx.db
+      .query(request.table)
+      .withIndex(index, rangeToQuery(range))
+      .order(order);
+    for await (const doc of query) {
+      if (page.length >= absoluteLimit) {
+        hasMore = true;
+        break;
+      }
+      page.push(doc);
+      indexKeys.push(getIndexKey(doc, indexFields));
+    }
+  }
```

Comment:
whoops good call.
i accidentally thought the `break;` would break out of the outer loop. to prevent such confusion, i'll change this to an early return.